### PR TITLE
[thread] add Thread stack lock held for output the unicast addresses

### LINKED ISSUE #33668: Missing Thread stack lock can lead to crashes during ConnectToNetwork on some Thread devices during commissioning
We observed that there can sometimes be, rarely but not never, crashes of Thread-based devices during commissioning that arises during OpenThread state change even logging on the Thread interface.

Example:

```
[29464765] <dbg> chip: LogV: [DL]OpenThread State Changed (Flags: 0x1100100d)
[29464978] <dbg> chip: LogV: [DL]   Device Role: DETACHED
[29465181] <dbg> chip: LogV: [DL]   Thread Unicast Addresses:
[29465401] <dbg> chip: LogV: [DL]        fdd0:876f:197f:0:b787:7e00:f5a1:8ed8/64 valid preferred
[29465727] <dbg> chip: LogV: [DL]        fe80::6886:bbc8:913c:e295/64 valid preferred
[29508887] <inf> chip: [DL]SRP Client was started, detected server: fdd0:876f:197f:0000:94f1:c03e:7c4d:4e9f
#### SRP registration complete.
[29508958] <dbg> chip: LogV: [DL]OpenThread State Changed (Flags: 0x00020d9d)
[29508992] <dbg> chip: LogV: [DL]   Device Role: CHILD
[29509019] <dbg> chip: LogV: [DL]   Network Name: NEST-PAN-2645
[29509152] <dbg> chip: LogV: [DL]   PAN Id: 0x2645
[29509346] <dbg> chip: LogV: [DL]   Extended PAN Id: 0x3E81DFCDA4579734
[29509590] <dbg> chip: LogV: [DL]   Channel: 13
[29509763] <dbg> chip: LogV: [DL]   Mesh Prefix: fdd0:876f:197f::/64
[29510001] <dbg> chip: LogV: [DL]   Partition Id: 0x837ED6E
[29510210] <dbg> chip: LogV: [DL]   Thread Unicast Addresses:
[29510430] <dbg> chip: LogV: [DL]        fd00:3105:a2e7:1:b622:7961:9ea3:6b27/64 valid preferred
#### UNEXPECTED RESTART BELOW, see timestamp returning to zero.
[00001191] <dbg> chip: LogV: [-]Starting pw_rpc server
```

During analysis, we found that [this line](https://github.com/project-chip/connectedhomeip/blob/194d91876a4ddd367619f02bdf763c45afaa4bcc/src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp#L250) calls a method with the following documentation:

```
/**
 * Log information related to a state change in the OpenThread stack.
 *
 * NB: This function *must* be called with the Thread stack lock held.
 */
void LogOpenThreadStateChange(otInstance * otInst, uint32_t flags)
```

However, at the call site, the Thread stack lock is missing.

### PR DESCRIPTION
**Pull request by ZhangLe2016**
Fixes #33668

### CODE REVIEW

**Comment by andy31415** [comment]
@ZhangLe2016 please sign the CLA

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp [hunk 1]
```diff
@@ -247,7 +247,9 @@ void GenericThreadStackManagerImpl_OpenThread<ImplClass>::_OnPlatformEvent(const
         }
 
 #if CHIP_DETAIL_LOGGING
+        Impl()->LockThreadStack();
```
Comment (tcarmelveilleux) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:250]: Please hold-off, @jmartinez-silabs is testing

Comment (tcarmelveilleux) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:250]: The comment at the top of the method `LogOpenThreadStateChange` about the lock is in the .cpp, but it should have been in the .h. This PR should also improve the documentation of the method.

Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:250]: @ZhangLe2016 @tcarmelveilleux I've tested on my end with silabs hardware. No issue was found. _OnPlatformEvent is always called by the chip task and LogOpenThreadStateChange uses direct thread stack API and does not try to lock the ThreadStack itself.

**Review by jmartinez-silabs — APPROVED** [review]

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 1 above)
Comment (ZhangLe2016) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:250]: Thanks a lot for your confirmation.

**Comment by jmartinez-silabs** [comment]
@ZhangLe2016 CLA not signed yet. Could you check if you completed the signing

**Review by tcarmelveilleux — APPROVED** [review]

**Pull request merged**